cloud-spanner-r2dbc/src/main/java/com/google/cloud/spanner/r2dbc/v2/SpannerClientLibraryConnection.java

```
@@ -51,7 +58,18 @@ public Publisher<Void> beginTransaction() {
 
   @Override
   public Publisher<Void> beginTransaction(TransactionDefinition definition) {
-    return Mono.error(new UnsupportedOperationException());
+    IsolationLevel isolationLevel = firstNonNull(definition.getAttribute(ISOLATION_LEVEL),
+        SERIALIZABLE);
+    return validateIsolation(isolationLevel)
+        .then(Mono.defer(() -> {
+          boolean isReadOnly = TRUE.equals(definition.getAttribute(READ_ONLY));
+          if (isReadOnly) {
+            TimestampBound timestampBound = firstNonNull(definition.getAttribute(TIMESTAMP_BOUND),
+                TimestampBound.strong());
+            return this.clientLibraryAdapter.beginReadonlyTransaction(timestampBound);
+          }
+          return this.clientLibraryAdapter.beginTransaction();
```

I think that we should throw an error if the user has created a `TransactionDefinition` for a read/write transaction with a read-only staleness. That is; if they have set a `TIMESTAMP_BOUND` option, but not set `READ_ONLY` to true.

---

Addressed in [4e41ba2](https://github.com/GoogleCloudPlatform/cloud-spanner-r2dbc/pull/738/commits/4e41ba2f4fc1c632c74b4ead8dacbcb7d2934961).